nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,159 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(
+        {%{shape: {m, k} = shape_q} = q_holder,
+         %{shape: {k, n} = shape_r, type: {_, num_bits} = type} = r_holder},
+        tensor,
+        opts
+      ) do
+    mode = opts[:mode]
+
+    r =
+      if mode == :reduced do
+        tensor[[0..(k - 1), 0..(n - 1)]]
+      else
+        tensor
+      end
+      |> Nx.as_type(type)
```

Comment:
Don't do the type conversion upfront if possible. Instead keep the type as is and it should transparently become the proper type as you do number_to_binary at the end.